File: src/AJKIOT.Shared/Models/IotDevice.cs
using AJKIOT.Shared.Models.DeviceFeatures;
using System.ComponentModel.DataAnnotations;
using System.Text.Json;

namespace AJKIOT.Shared.Models
{
    public class IotDevice
    {
        [Key]
        public int Id { get; set; }
        public string OwnerId { get; set; } = string.Empty;
        public string DeviceType { get; set; } = string.Empty;
        public string DeviceName { get; set; } = string.Empty;
        public string DeviceFeaturesJson { get; set; } = string.Empty;
        public string DeviceScheduleJson { get; set; } = string.Empty;


        public void SetFeatures(IEnumerable<DeviceFeature> deviceFeatures)
        {
            if (deviceFeatures != null)
                DeviceFeaturesJson = JsonSerializer.Serialize(deviceFeatures);
        }


        public IEnumerable<DeviceFeature> GetFeatures()
        {
            try
            {
                if (!string.IsNullOrEmpty(DeviceFeaturesJson))
                    return JsonSerializer.Deserialize<IEnumerable<DeviceFeature>>(DeviceFeaturesJson)!;
                else
                    return new List<DeviceFeature>();
            }
            catch
            {
                return new List<DeviceFeature>();
            }
        }

        public void SetSchedule(IEnumerable<DailyScheduleEntry> dailyScheduleEntries)
        {
            if (dailyScheduleEntries != null)
                DeviceScheduleJson = JsonSerializer.Serialize(dailyScheduleEntries);
        }

        public IEnumerable<DailyScheduleEntry> GetSchedule()
        {
            try
            {
                if (!string.IsNullOrEmpty(DeviceScheduleJson))
                    return JsonSerializer.Deserialize<IEnumerable<DailyScheduleEntry>>(DeviceScheduleJson)!;
                else
                    return new List<DailyScheduleEntry>();
            }
            catch
            {
                return new List<DailyScheduleEntry>();
            }
        }
    }
}

File: src/ajk_iot_mobile/lib/models/iot_device.dart
import 'dart:convert';

import 'package:ajk_iot_mobile/models/daily_schedule_entry.dart';
import 'package:ajk_iot_mobile/models/device_feature.dart';

class IotDevice {
  int id;
  String ownerId;
  String deviceType;
  String deviceName;
  String deviceFeaturesJson;
  String deviceScheduleJson;

  IotDevice({
    required this.id,
    required this.ownerId,
    required this.deviceType,
    required this.deviceName,
    required this.deviceFeaturesJson,
    required this.deviceScheduleJson,
  });

  bool isScheduleAvailable() {
    switch (deviceType) {
      case 'Switch':
        return true;
      case 'OpenTimer':
        return false;
      default:
        return false;
    }
  }

  void setFeatures(List<DeviceFeature> deviceFeatures) {
    deviceFeaturesJson = jsonEncode(deviceFeatures.map((f) => f.toJson()).toList());
  }

  List<DeviceFeature> getFeatures() {
    return jsonDecode(deviceFeaturesJson).map<DeviceFeature>((f) => DeviceFeature.fromJson(f)).toList();
  }

  void setSchedule(List<DailyScheduleEntry> dailyScheduleEntries) {
    if (isScheduleAvailable()) {
      deviceScheduleJson = jsonEncode(dailyScheduleEntries.map((e) => e.toJson()).toList());
    }
  }

  List<DailyScheduleEntry> getSchedule() {
    if (isScheduleAvailable() && deviceScheduleJson.isNotEmpty) {
      return jsonDecode(deviceScheduleJson).map<DailyScheduleEntry>((e) => DailyScheduleEntry.fromJson(e)).toList();
    } else {
      return [];
    }
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'ownerId': ownerId,
    'deviceType': deviceType,
    'deviceName': deviceName,
    'deviceFeaturesJson': deviceFeaturesJson,
    'deviceScheduleJson': deviceScheduleJson,
  };

  static IotDevice fromJson(Map<String, dynamic> json) => IotDevice(
    id: json['id'],
    ownerId: json['ownerId'],
    deviceType: json['deviceType'],
    deviceName: json['deviceName'],
    deviceFeaturesJson: json['deviceFeaturesJson'],
    deviceScheduleJson: json['deviceScheduleJson'],
  );
}
